### PR TITLE
feat: add retry mechanism

### PR DESCRIPTION
**Pull request by Salka1988**
Closes https://github.com/FuelLabs/fuels-rs/issues/1020. 

This code is an implementation of a retry mechanism that allows a function to be retried a certain number of times with a specified interval between attempts.

`RetryConfig`: This is a configuration struct that holds the retry configuration. It contains the following fields:

- max_attempts: The maximum number of times the action should be retried.
- interval:  Determines the backoff strategy and timing to be used during retry attempts. It allows you to configure how much time a system should wait before retrying an action after a failure. 

`retry<Fut, T, ShouldRetry>`: This is the main function responsible for implementing the retry logic.

This function takes several parameters:

- action: A closure that performs the action you want to retry. It returns a Future of the result.
- retry_config: A reference to a `RetryConfig` struct that specifies the retry behavior.
- should_retry: A closure that takes a reference to the result of the action and returns a boolean indicating whether the action should be retried based on the result.

Aditional Work:
I've separated sending a transaction from getting back its value:
```rust
  let response = contract_methods.get(1, 2).submit().await?;
  let value = response.response().await?;
```

Added `RetryableClient` a wrapper around the `FuelClient` that provides a convenient way to interact with a remote service. It also enables automatic retry functionality in scenarios where network issues may occur.

```rust
#[derive(Debug, Clone)]
pub(crate) struct RetryableClient {
    client: FuelClient,
    url: String,
    retry_config: RetryConfig,
}
```

### Checklist
- [x] I have linked to any relevant issues.
- [x] I have updated the documentation.
- [x] I have added necessary labels.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Review by Br1ght0ne — CHANGES_REQUESTED** [review]
When designing a retry mechanism for anything with a notion of “network call” in some way, I like implementing a delay to allow the network to self-regulate its throughput and often requiring less attempts overall.

I usually go for [exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff) - delays start off small and ramp up steadily. There are crates like [`backon`](https://lib.rs/crates/backon) with full support of sync/async functions.

Alternatives:
- hand-roll our own delayed retrying (only if adding a dependency is a problem)
- skip having delays altogether (I hereby request your comments on it)

**Comment by Salka1988** [comment]
> Great UX and overall approach! I have a small consideration regarding the logic itself.
> 
> When designing a retry mechanism for anything with a notion of “network call” in some way, I like implementing a delay to allow the network to self-regulate its throughput when congested and often requiring less attempts overall.
> 
> I usually go for [exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff) - delays start off small and ramp up steadily. There are crates like [`backon`](https://lib.rs/crates/backon) and [`backoff`](https://lib.rs/crates/backoff) with full support of sync/async functions.
> 
> Extra considerations:
> 
> * what are acceptable minimum/maximum delays for our use case?
> 
> Alternatives:
> 
> * hand-roll our own delayed retrying (only if adding a dependency is a problem)
> * skip having delays altogether (I hereby request your comments on it)

I will use one of the crates you recommended. If that's a problem, I'll implement `exponential backoff'.

**Review by segfault-magnet — COMMENTED** [review]
Not all errors benefit from retrying. Usually, you want to wait out a transient error (network or some other kind) and not actually offer a fully-fledged retry mechanism to the user especially when retrying again so fast.

For example, if the tx is out of gas it will be reverted and you'll retry again most probably changing nothing (and more importantly - wasting gas).

**Comment by segfault-magnet** [comment]
> > Not all errors benefit from retrying. Usually, you want to wait out a transient error (network or some other kind) and not actually offer a fully-fledged retry mechanism to the user especially when retrying again so fast.
> > For example, if the tx is out of gas it will be reverted and you'll retry again most probably changing nothing (and more importantly - wasting gas).
> 
> You are right, I understand you but will we waste gas if we fail?

Yep. You lose the gas used to execute the contract up until the point of reverting. The rest will be refunded if you provided an adequate change output.

**Comment by segfault-magnet** [comment]
> > Not all errors benefit from retrying. Usually, you want to wait out a transient error (network or some other kind) and not actually offer a fully-fledged retry mechanism to the user especially when retrying again so fast.
> > For example, if the tx is out of gas it will be reverted and you'll retry again most probably changing nothing (and more importantly - wasting gas).
> 
> You are right, I understand you but will we waste gas if we fail?

Yep. You lose the gas used to execute the contract up until the point of reverting. The rest will be refunded if you provided an adequate change output.

**Comment by Br1ght0ne** [comment]
Thanks for your consideration @Salka1988! I agree with @segfault-magnet - saving on gas is important, and a proper `max_attempts` should be set to avoid excessive retrying. I dismissed my blocking review, and will approve whatever logic we agree on (if the code is correct 😉)

**Comment by Salka1988** [comment]
> > > Not all errors benefit from retrying. Usually, you want to wait out a transient error (network or some other kind) and not actually offer a fully-fledged retry mechanism to the user especially when retrying again so fast.
> > > For example, if the tx is out of gas it will be reverted and you'll retry again most probably changing nothing (and more importantly - wasting gas).
> > 
> > 
> > You are right, I understand you but will we waste gas if we fail?
> 
> Yep. You lose the gas used to execute the contract up until the point of reverting. The rest will be refunded if you provided an adequate change output.

Yea! I forgot about `simulate()`

File: packages/fuels-programs/src/contract.rs
Comment (iqdecay) [inline packages/fuels-programs/src/contract.rs:482]: these functions should have docstrings

File: packages/fuels-test-helpers/src/node.rs
Comment (iqdecay) [inline packages/fuels-test-helpers/src/node.rs:63]: call it `chain_config` I think, we are not using `conf` anywhere in the codebase.

File: packages/fuels-programs/src/contract.rs
Comment (iqdecay) [inline packages/fuels-programs/src/contract.rs:722]: why not just `call_with_retry`? For `MultiContractCallHandler` we use the `call` method, the same as `ContractCallHandler` so I don't see why we would have different names here.

File: packages/fuels-programs/src/contract.rs
Comment (iqdecay) [inline packages/fuels-programs/src/contract.rs:493]: I think we're better off matching with the `ClientRequestError` variant rather than a string description that can change.

File: packages/fuels/tests/contracts.rs
Comment (iqdecay) [inline packages/fuels/tests/contracts.rs:1]: why are those tests commented out?

**Review by iqdecay — CHANGES_REQUESTED** [review]
looking good, left a few questions.

File: packages/fuels-accounts/src/provider.rs
Comment (segfault-magnet) [inline packages/fuels-accounts/src/provider.rs:366]:
```suggestion
            .await?
```

File: packages/fuels-programs/src/contract.rs
Comment (segfault-magnet) [reply packages/fuels-programs/src/contract.rs:482]: I'd also create a `RetryConfig` with some defaults and place the now hardcoded interval inside of there alongside the max attempts.

File: packages/fuels-programs/src/contract.rs
Comment (segfault-magnet) [reply packages/fuels-programs/src/contract.rs:493]: Yep. I'd leave a match statement that retries only if the provider error was network related. 

Not sure how fine-tuned you can make that. For example, schema incompatibility between our gql client and the server might also appear as the same type of error (since fuel gql client uses `anyhow`). I suggest digging a bit to see whether you can make any sort of robust claim as to the transitive nature of the error.

You can always send a PR over to the client for more library-friendly errors.

File: packages/fuels-programs/src/contract.rs
Comment (segfault-magnet) [reply packages/fuels-programs/src/contract.rs:722]: Probably a paste-ing mishap.

File: packages/fuels-programs/src/contract.rs
Comment (Salka1988) [reply packages/fuels-programs/src/contract.rs:482]: I did that, but I tried to add some functionality like  `on_retry: Option<Box<dyn Fn(usize) -> std::pin::Pin<Box<dyn std::future::Future<Output=()>>> + Send + Sync>>` function that would allow us to change certain contract parameters in a certain repetition or similar. I added this primarily to test certain things, but everything got a little complicated.:goberserk:
 Does it even make sense for me to add this functionality

File: packages/fuels-test-helpers/src/node.rs
Comment (Salka1988) [reply packages/fuels-test-helpers/src/node.rs:63]: This was added for testing purposes. I don't think it will be necessary.

File: packages/fuels-programs/src/contract.rs
Comment (Salka1988) [reply packages/fuels-programs/src/contract.rs:493]: You're right, but this is currently the "only" way to do matching. I should talk to the client team.

**Comment by Salka1988** [comment]
We should wait for #1073 to merge before we continue with the implementation.

File: packages/fuels-accounts/src/provider.rs
Comment (MujkicA) [inline packages/fuels-accounts/src/provider.rs:147]: Nit: it seems a bit wordy, maybe `send_transaction_and_await`?

File: packages/fuels-programs/src/contract.rs
Comment (MujkicA) [inline packages/fuels-programs/src/contract.rs:618]: If you go with the suggestion about send/submit in the provider, we should align this to say send as well

File: packages/fuels-test-helpers/src/node.rs
Comment (MujkicA) [reply packages/fuels-test-helpers/src/node.rs:64]: Reminder to remove it

File: docs/src/running-scripts.md
Comment (Salka1988) [reply docs/src/running-scripts.md:41]: mistake!

File: packages/fuels-programs/src/contract.rs
Comment (Salka1988) [reply packages/fuels-programs/src/contract.rs:472]: I needed this until @hal3e  joined one of his PRs. It stayed like this because it didn't cause me a problem, so I forgot about it. Now we can do as you said. `Clone`

File: packages/fuels-programs/src/retry.rs
Comment (Salka1988) [reply packages/fuels-programs/src/retry.rs:48]: Nice! Thanx :)

Comment (MujkicA) [reply packages/fuels-programs/src/retry.rs:48]: What should be the expected behaviour if someone enters 0?

File: packages/fuels-programs/src/submit_response.rs
Comment (MujkicA) [inline packages/fuels-programs/src/submit_response.rs:85]: Why do we need this indirection

File: packages/fuels-programs/src/retry.rs
Comment (MujkicA) [inline packages/fuels-programs/src/retry.rs:163]: If the count starts at 0 here, and the strategy is `Linear`, the first retry will occur with no delay. Is this expected of linear backoff?

File: packages/fuels-programs/src/submit_response.rs
Comment (Salka1988) [reply packages/fuels-programs/src/submit_response.rs:85]: We dont need it.

**Review by Br1ght0ne — APPROVED** [review]
Very thorough and well-designed PR. LGTM!

File: packages/fuels-programs/src/retry.rs
Comment (MujkicA) [inline packages/fuels-programs/src/retry.rs:167]: Can we have something like if attempts == 0 -> Error on the top of the method.
IMO it's more readable that way

Comment (Salka1988) [reply packages/fuels-programs/src/retry.rs:167]: Ok

File: packages/fuels-programs/src/retry.rs
Comment (hal3e) [inline packages/fuels-programs/src/retry.rs:310]: maybe we can extract this to a `tx_id` variable and later just use `assert_eq!(ok, tx_id);`
```suggestion
            let tx_id = TxId::from_str(
                "0x98f01c73c2062b55bba70966917a0839995e86abfadfff24534262d1c8b7a64e",
            );
            let values = Mutex::new(vec![
                Ok(tx_id),
```

File: packages/fuels-programs/src/retry.rs
Comment (hal3e) [inline packages/fuels-programs/src/retry.rs:333]:
```suggestion
            assert_eq!(ok, tx_id);
```

File: packages/fuels-programs/src/script_calls.rs
Comment (hal3e) [inline packages/fuels-programs/src/script_calls.rs:283]: take it if you like it:
```suggestion
        self.cached_tx_id = match self.retry_config.max_attempts {
            0 => provider.send_transaction(tx).await,
            _ => {
                retry(
                    || async { provider.send_transaction(tx.clone()).await },
                    &self.retry_config,
                    should_retry_fn,
                )
                .await
            }
        }
        .map(Some)?;
```

**Review by hal3e — CHANGES_REQUESTED** [review]
This is such a great improvement! I love the fact that we can configure how we retry our calls. The code and test are awesome! Great work!

I have mad many suggestions but they are actually just small things that, if you decide, can be added very easily. Additionally, I have added all suggestions on all lines because I wanted to make it easier for you to accept them if you decided. I did this because I was already making the changes and checking if everything works so instead of making a new PR/commit I have just added them as suggestions.

**Review by Br1ght0ne — APPROVED** [review]

**Review by segfault-magnet — DISMISSED** [review]
Reviewed and gave feedback live with @Salka1988 .

**Comment by Salka1988** [comment]
> Reviewed and gave feedback live with @Salka1988 .

@segfault-magnet  creativity and expertise truly elevated the outcome. Thanx!

**Review by Br1ght0ne — APPROVED** [review]

**Review by hal3e — APPROVED** [review]
Awesome work!

**Pull request merged**
